src/AggregateFunctions/AggregateFunctionGroupArrayIntersect.cpp

```diff
@@ -0,0 +1,466 @@
+#include <AggregateFunctions/FactoryHelpers.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDate32.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeDateTime64.h>
+
+
+namespace DB
+{
+struct Settings;
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+struct Settings;
+
+
+template <typename T>
+struct AggregateFunctionGroupArrayIntersectData
+{
+    using Map = HashMap<T, UInt64>;
+    Map value;
+    UInt64 version = 0;
+};
+
+
+/// Puts all values to the hash set. Returns an array of unique values. Implemented for numeric types.
+template <typename T>
+class AggregateFunctionGroupArrayIntersect
+    : public IAggregateFunctionDataHelper<AggregateFunctionGroupArrayIntersectData<T>, AggregateFunctionGroupArrayIntersect<T>>
+{
+
+private:
+    using State = AggregateFunctionGroupArrayIntersectData<T>;
+
+public:
+    AggregateFunctionGroupArrayIntersect(const DataTypePtr & argument_type, const Array & parameters_)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupArrayIntersectData<T>,
+          AggregateFunctionGroupArrayIntersect<T>>({argument_type}, parameters_, std::make_shared<DataTypeArray>(argument_type)) {}
+
+    AggregateFunctionGroupArrayIntersect(const DataTypePtr & argument_type, const Array & parameters_, const DataTypePtr & result_type_)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupArrayIntersectData<T>,
+          AggregateFunctionGroupArrayIntersect<T>>({argument_type}, parameters_, result_type_) {}
+
+    String getName() const override { return "GroupArrayIntersect"; }
+
+    bool allocatesMemoryInArena() const override { return false; }
+
+    void add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena *) const override
+    {
+        auto & version = this->data(place).version;
+        auto & map = this->data(place).value;
+
+        const auto data_column = assert_cast<const ColumnArray &>(*columns[0]).getDataPtr();
+        const auto & offsets = assert_cast<const ColumnArray &>(*columns[0]).getOffsets();
+        const size_t offset = offsets[static_cast<ssize_t>(row_num) - 1];
+        const auto arr_size = offsets[row_num] - offset;
+
+        ++version;
+        if (version == 1)
+        {
+            for (size_t i = 0; i < arr_size; ++i)
+                map[static_cast<T>((*data_column)[offset + i].get<T>())] = version;
+        }
+        else if (map.size() > 0)
+        {
+            typename State::Map new_map;
+            for (size_t i = 0; i < arr_size; ++i)
+            {
+                typename State::Map::LookupResult value = map.find(static_cast<T>((*data_column)[offset + i].get<T>()));
+                if (value != nullptr && value->getMapped() == version - 1)
+                    new_map[static_cast<T>((*data_column)[offset + i].get<T>())] = version;
+            }
+            map = std::move(new_map);
```

Comment:
Can `value->getMapped()` store values other than `version - 1`?
The `value->getMapped() == version - 1` check seems redundant since obsolete values are not preserved in the `new_map`.

If so, can `AggregateFunctionGroupArrayIntersectData` be based on a hash set instead of the hash map, omitting the `version` value?